### PR TITLE
[rush] Fix 1347 by resolving paths to local projects by referring to the shrinkwrap

### PR DESCRIPTION
**Pull request by sachinjoseph**
`rush link` was failing on pnpm 3+ with the changes in shrinkwrap format with regard to peer dependencies. Rush now resolves paths to local projects accurately by referring to the shrinkwrap rather than figuring out the path on its own. See https://github.com/microsoft/web-build-tools/issues/1347.

### CODE REVIEW

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 1]
```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
     // appropriate. This folder is usually something like:
     // C:\{uri-encoed-path-to-tgz}\node_modules\{package-name}
 
+    // e.g.:
+    //   file:projects/bentleyjs-core.tgz
+    //   file:projects/build-tools.tgz_dc21d88642e18a947127a751e00b020a
+    //   file:projects/imodel-from-geojson.tgz_request@2.88.0
+    const topLevelDependencyVersion: string | undefined =
+      pnpmShrinkwrapFile!.getTopLevelDependencyVersion(project.tempProjectName);
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: This will throw if `pnpmShrinkwrapFile` is undefined. Either drop the `undefined` from the type, or perform a defensive check.

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 2]
```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
...
+    } else {
+      relativePathToTgzFile = topLevelDependencyVersion.slice('file:'.length);
+    }
+
     // e.g.: C:\wbt\common\temp\projects\api-documenter.tgz
-    const pathToTgzFile: string = path.join(
-      this._rushConfiguration.commonTempFolder,
-      'projects',
-      `${unscopedTempProjectName}.tgz`);
+    const absolutePathToTgzFile: string = path.join(this._rushConfiguration.commonTempFolder, relativePathToTgzFile);
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: Use `path.resolve`:

```suggestion
    const absolutePathToTgzFile: string = path.resolve(this._rushConfiguration.commonTempFolder, relativePathToTgzFile);
```

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts [hunk 3]
```diff
@@ -146,18 +147,25 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
   }
 
   /**
-   * Serializes the PNPM Shrinkwrap file
+   * Gets the version number from the list of top-level dependencies in the "dependencies" section
+   * of the shrinkwrap file
    */
-  protected serialize(): string {
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: Can you move these back to where they were in this file to minimize the code change?

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts [hunk 4]
```diff
@@ -146,18 +147,25 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
...
   }
 
   /**
-   * Gets the version number from the list of top-level dependencies in the "dependencies" section
-   * of the shrinkwrap file
+   * Gets the id of the specified package
    */
-  protected getTopLevelDependencyVersion(dependencyName: string): string | undefined {
-    return BaseShrinkwrapFile.tryGetValue(this._shrinkwrapJson.dependencies, dependencyName);
+  public getPackageId(packageName: string): string | undefined {
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: What is the "ID?" Can you provide some better documentation and/or a better function name?

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 5]
```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
...
+    //   file:projects/imodel-from-geojson.tgz_request@2.88.0
+    const topLevelDependencyVersion: string | undefined =
+      pnpmShrinkwrapFile!.getTopLevelDependencyVersion(project.tempProjectName);
+
+    if (!topLevelDependencyVersion) {
+      throw new InternalError(`Cannot find top level dependency for "${project.tempProjectName}"` +
+        ` in shrinkwrap.`);
+    }
+
+    const packageId: string | undefined = pnpmShrinkwrapFile!.getPackageId(topLevelDependencyVersion);
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: Why wouldn't we get a package ID? Can you provide documentation on when/why that can happen?

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 6]
```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
...
+
+    // e.g.: projects\api-documenter.tgz
+    let relativePathToTgzFile: string;
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: You can clean this up with a conditional:

```suggestion
    const relativePathToTgzFile: string = packageId
     ? packageId.slice('file:'.length)
     :  topLevelDependencyVersion.slice('file:'.length);
```

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 7]
```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
...
+    const absolutePathToTgzFile: string = path.join(this._rushConfiguration.commonTempFolder, relativePathToTgzFile);
 
-    // e.g.: C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz
-    const escapedPathToTgzFile: string = uriEncode(Text.replaceAll(pathToTgzFile, path.sep, '/'));
+    // e.g.:
+    //   C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz
+    //   C%3A%2Fdev%2Fimodeljs%2Fimodeljs%2Fcommon%2Ftemp%2Fprojects%2Fpresentation-integration-tests.tgz_jsdom@11.12.0
+    //   C%3A%2Fdev%2Fimodeljs%2Fimodeljs%2Fcommon%2Ftemp%2Fprojects%2Fbuild-tools.tgz_2a665c89609864b4e75bc5365d7f8f56
+    const dirNameInLocal: string = uriEncode(Text.replaceAll(absolutePathToTgzFile, path.sep, '/')) +
+      (packageId && packageId.length < topLevelDependencyVersion.length ?
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: This is somewhat hard to read. Can you split the `packageId && ...` stuff into an intermediate variable?

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 8]
```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
...
-      `${unscopedTempProjectName}.tgz`);
+    const absolutePathToTgzFile: string = path.join(this._rushConfiguration.commonTempFolder, relativePathToTgzFile);
 
-    // e.g.: C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz
-    const escapedPathToTgzFile: string = uriEncode(Text.replaceAll(pathToTgzFile, path.sep, '/'));
+    // e.g.:
+    //   C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz
+    //   C%3A%2Fdev%2Fimodeljs%2Fimodeljs%2Fcommon%2Ftemp%2Fprojects%2Fpresentation-integration-tests.tgz_jsdom@11.12.0
+    //   C%3A%2Fdev%2Fimodeljs%2Fimodeljs%2Fcommon%2Ftemp%2Fprojects%2Fbuild-tools.tgz_2a665c89609864b4e75bc5365d7f8f56
+    const dirNameInLocal: string = uriEncode(Text.replaceAll(absolutePathToTgzFile, path.sep, '/')) +
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: We generally use the word `folder` instead of `dir` or `directory`.

Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: Is this a folder name? Or is it a full filename path?

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 3 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: Sure

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts (hunk 7 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: Will do

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts (hunk 8 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: Folder name. Examples are given in the comments.

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 4 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: Will do.

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 3 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: I want `getTopLevelDependencyVersion` exposed, so I had to move it above `protected` methods.

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 4 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: I was using the id field in the shrinkwrap to get the path to a project tarball, but found that there is a field called `tarball` which is precisely what I need.

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts [hunk 9]
```diff
@@ -146,20 +146,28 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
...
 
   /**
-   * Serializes the PNPM Shrinkwrap file
+   * Gets the path to the tarball file if the package is a tarball,
+   * otherwise returns undefined. Example: file:projects/build-tools.tgz
    */
-  protected serialize(): string {
-    return yaml.safeDump(this._shrinkwrapJson, SHRINKWRAP_YAML_FORMAT);
+  public getTarballPath(packageName: string): string | undefined {
+    return this._shrinkwrapJson.packages[packageName].resolution.tarball;
```
Comment (iclanton) [inline apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: We should have a defensive check around `this._shrinkwrapJson.packages[packageName]`. It should probably return `undefined` if `this._shrinkwrapJson.packages[packageName]` is `undefined` instead of throwing.

**Comment by iclanton** [comment]
Make sure you test this pretty extensively.

File: apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts (hunk 9 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts]: Hmm, makes sense.

**Comment by sachinjoseph** [comment]
> 
> 
> Make sure you test this pretty extensively.

@iclanton I tested this on wbt, sp-client, and the repro repo in #1347, https://github.com/imodeljs/imodeljs, on pnpm 2 and pnpm 3. Do you suggest more test cases? I am not completely familiar with the feature so it's possible that I am missing some edge cases.

**Comment by iclanton** [comment]
I think you're good.
